scipy/sparse/_coo.py

```diff
@@ -722,16 +778,490 @@ def _matmul_multivector(self, other):
+                other.shape
+            except AttributeError:
+                other = o_array
+
+        if self.ndim < 3 and (np.isscalar(other) or other.ndim<3):
+            return _spbase.dot(self, other)
+        if isdense(other):
+            return self._dense_dot(other)
+        else:
+            if not (isinstance(self, (coo_array, coo_matrix)) or np.isscalar(self)):
+                raise TypeError("array must be in COO format")
+            if not (isinstance(other, (coo_array, coo_matrix)) or np.isscalar(other)):
+                raise TypeError("input must be a COO matrix/array or a scalar")
+
+            # Handle scalar multiplication
+            if np.isscalar(other):
+                return self * other
+
+            # Handle inner product of vectors (1-D arrays)
+            if self.ndim == 1 and other.ndim == 1:
+                if self.shape[0] != other.shape[0]:
+                    raise ValueError(f"shapes {self.shape} and {other.shape}"
+                                     " are not aligned for inner product")
+                return self @ other
+            
+            # Handle matrix multiplication (2-D arrays)
+            if self.ndim == 2 and other.ndim == 2:
+                if self.shape[1] != other.shape[0]:
+                    raise ValueError(f"shapes {self.shape} and {other.shape}"
+                                     " are not aligned for matmul")
+                return self @ other
+            
+            return self._sparse_dot(other)
+
+    
+    def _sparse_dot(self, other):
+        self_is_1d = False
+        other_is_1d = False
+
+        # reshape to 2-D if self or other is 1-D
+        if self.ndim == 1:
+            self = self.reshape((1, self.shape[0])) # prepend 1 to shape
+            self_is_1d = True
+
+        if other.ndim == 1:
+            other = other.reshape((other.shape[0], 1)) # append 1 to shape
+            other_is_1d = True
+
+        if self.shape[-1] != other.shape[-2]:
+                raise ValueError(f"shapes {self.shape} and {other.shape}"
+                                 " are not aligned for n-D dot")
+        
+        # Prepare the tensors for dot operation
+        # Ravel non-reduced axes coordinates
+        self_raveled_coords = _ravel_non_reduced_axes(self.coords,
+                                                      self.shape, [self.ndim-1])
+        other_raveled_coords = _ravel_non_reduced_axes(other.coords,
+                                                       other.shape, [other.ndim-2])
+
+        # Get the shape of the non-reduced axes
+        og_shape_self = self.shape[:-1]
+        og_shape_other = other.shape[:-2] + other.shape[-1:]
+        
+        # Create 2D coords arrays
+        ravel_coords_shape_self = (math.prod(og_shape_self), self.shape[-1])
+
+        ravel_coords_shape_other = (other.shape[-2], math.prod(og_shape_other))
+        
+        
+        self_2d_coords = np.vstack((self_raveled_coords, self.coords[-1]))
```

Comment:
vstack will cause a copy here. Let's keep things in tuple format:

self_2d_coords = self_raveled_coords + (self.coords[-1],)

Reply:
I think using `+` performed addition of the arrays, so I made the following change instead - 96af7953f727edc4325a6f02d4c53b09ddfd7295

Reply:
The addition works when we stick with coords being a tuple of 1D arrays. So, e.g. in this case `self_raveled_coords` will need to be a tuple of 1D arrays as well. Its often easier to avoid copying when we manipulate these tuples than when using a single 2D array. In this case you do need to change how the helper functions do it too though.

Reply:
Done at https://github.com/scipy/scipy/pull/21435/commits/45ad0b573825b576665b2435ad23eccc23ac5b41! Also spotted a small bug in a helper function so now that's fixed too :)